astropy/nddata/utils.py

```diff
@@ -242,7 +242,11 @@ def extract_array(array_large, shape, position, mode='partial',
 
     # Extracting on the edges is presumably a rare case, so treat special here
     if (extracted_array.shape != shape) and (mode == 'partial'):
-        extracted_array = np.zeros(shape, dtype=array_large.dtype)
+        if np.isnan(fill_value):
```

Comment:
Do you think there needs to be info or warning when array type is changed like this?

Reply:
Perhaps, but we weren't doing it before.

Reply:
We could also add a note in the docstring.

Reply:
Yeah, maybe in the docstring then.

Reply:
The docstring says "``fill_value`` must have the same ``dtype`` as the ``array_large`` array." which could be improved.
But indeed if the output array dtype is changed because of the fill value type, I think a warning would be appropriate.
Another option would be to have another default fill value for integers, I think 0 would be fine (by default).

Reply:
Not `-999`? 😏

Reply:
OK, I'll add the warning.  I think that's better than having different default `fill_values` based on the input `dtype`.

Reply:
It would not work for int8 or unsigned int ;). 
And for a function that is extracting a piece of an array, I think padding with 0 by default is reasonable. At least it's less surprising that getting a float array if you give it some integer type, imo.

Reply:
Using `0` as a default sentinel value for "no data" can be surprising too.  The user can always specify `fill_value=0` if that is what they really want.

Having different defaults would also be surprising to get a different result based only on the input dype.  I can see users getting confused by this:

```python
>>> extract_array(np.ones((5, 5)), (3, 3), (0, 0))
array([[nan, nan, nan],
       [nan,  1.,  1.],
       [nan,  1.,  1.]])

>>> extract_array(np.ones((5, 5), dtype=int), (3, 3), (0, 0))
array([[0,  0,  0],
       [0,  1,  1],
       [0,  1,  1]])
```

Reply:
I know -- masked arrays! jk :)

Reply:
I added the warning and updated the docstring.